Add MPP support to LSPS2

johncantrell97 opened the pull request:
This should be based on top of #20 but I can't open a PR on top of a PR in my repo over here.  So if reviewing this please just look at the last commit.  Once #20 is merged I will rebase this PR.

When a payment is made that initiates a JIT channel open it is possible the payer uses MPP. If `payment_size_msat` is specified by client then the LSP needs to allow for MPP to be used.

This means when an HTLC is intercepted we need to collect the payment parts and only open the channel once the total amount we collected is enough to cover the original payment_size.

This PR introduces a Vec of `InterceptedHTLC`'s into the OutboundJITChannelState to track these.  There's a new `OutboundJITChannelState::PendingPaymentReceived` state that can be entered if MPP is supported and the first htlc_intercepted is not large enough to cover the requested payment_size.

It is possible to skip this state if MPP is not supported or the first payment is large enough.

The spec does not specify how the LSP should extract the fee from the collection of HTLCs received as part of the MPP.  In this pass I iterate over all parts and greedily take as much of the fee I can from each HTLC until the total fee amount has been accounted for.  I could imagine another strategy that tries to minimize the number of HTLCs that are used to pay the fee (sorting from largest to smallest and performing same strategy but I'm not sure it matters.

It's also unclear to me whether batch forwarding all of the HTLCs at once could run into problems in practice.

johncantrell97 commented on src/jit_channel/channel_manager.rs:
```diff
@@ -216,20 +228,68 @@ impl OutboundJITChannelState {
+				let amt_to_forward_msat = payment_size_msat - opening_fee_msat;
+
+				if htlc.expected_outbound_amount_msat >= payment_size_msat
+					&& amt_to_forward_msat > 0
```
This should probably be something like `amt_to_forward_msat >= min_htlc_msat` but there's no channel yet so we don't know what the minimum htlc amount our peer will set on the channel.  I don't think the spec mentions anything about this.

tnull commented:
This needs rebase now that lightningdevkit/lightning-liquidity#20 was merged.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -192,13 +198,19 @@ enum OutboundJITChannelState {
 		payment_size_msat: Option<u64>,
 		opening_fee_params: OpeningFeeParams,
 	},
+	PendingPaymentReceived {
```
The name is a bit ambiguous as it could be read in the sense that we *did* receive the pending payment, which is exactly the opposite of what the state is for. Maybe `PendingPaymentReceipt`?

johncantrell97 replied:
Yeah good point.  I think even can/should get rid of the first state of InvoiceParametersGenerated because really soon as that is given out you are already awaiting payment with an empty vec of htlcs received.

johncantrell97 replied:
well the state is really supposed to be that we received _part_ of the payment.  it only transitions here once the first htlc is received.

i wonder though if we should just get rid of the InvoiceParametersGenerated state and just immediately enter some kind of AwaitingPayment state that holds the htlcs and initializes with empty vec.

johncantrell97 replied:
maybe it's beneficial to immediately split into one of two states whether or not MPP is supported (based on presence of the payment_size_msat).

This probably simplifies the logic when an htlc arrives. It allows for the fast failure of non-mpp supported payments that are < required amount.

tnull replied:
Yeah, that could be a solution. I'll leave it up to you. I also commented something similar regarding simplifying the states below. Here my main comment was that the naming is ambiguous and we should at least rename the state.

johncantrell97 replied:
I ended up removing both states and now the initial state of an outbound channel is `AwaitingPayment`. If mpp is enabled it stays there until receives enough to cover payment_size+fee.  If mpp not enabled then after the first htlc is received, if it's not enough it fails immediately.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -216,20 +228,68 @@ impl OutboundJITChannelState {
+				let supports_mpp = payment_size_msat.is_some();
+				let payment_size_msat =
+					payment_size_msat.unwrap_or(htlc.expected_outbound_amount_msat);
+				let opening_fee_msat = compute_opening_fee(payment_size_msat, opening_fee_params.min_fee_msat, opening_fee_params.proportional).ok_or(ChannelStateError(format!("Could not compute valid opening fee with min_fee_msat = {}, proportional = {}, and expected_outbound_amount_msat = {}", opening_fee_params.min_fee_msat, opening_fee_params.proportional, htlc.expected_outbound_amount_msat)))?;
```
This line is more than 400 characters wide. Can we please break it up?

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -216,20 +228,68 @@ impl OutboundJITChannelState {
+
+				let total_expected_outbound_amount_msat = htlcs
+					.iter()
+					.fold(0, |total_msat, htlc| total_msat + htlc.expected_outbound_amount_msat);
```
nit: Could write this a tad simpler:
```suggestion
					.map(|htlc| htlc.expected_outbound_amount_msat).sum();
```

johncantrell97 replied:
Easier to read but probably _slightly_ less efficient unless the compiler does some magic to avoid two loops? I'll go with easier to read regardless because the # of htlcs is tiny.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -216,20 +228,68 @@ impl OutboundJITChannelState {
+					.iter()
+					.fold(0, |total_msat, htlc| total_msat + htlc.expected_outbound_amount_msat);
+
+				if total_expected_outbound_amount_msat >= *payment_size_msat {
```
This is the wrong way around, no? If our total HTLC values are still *smaller* than the payment size, we remain  in `PendingPaymentReceived`?

johncantrell97 replied:
yes seems backwards

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -216,20 +228,68 @@ impl OutboundJITChannelState {
+					}
+				}
+			}
+			OutboundJITChannelState::PendingPaymentReceived {
```
I wonder whether it would make sense to have this be in the default path, even when we don't support MPP? We could have our logic discern between MPP and non-MPP requests and just always check the necessary parameters in this state? If they are met, we advance to `PendingChannelOpen`, if not, we timeout after a while?

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -625,25 +683,17 @@ where
 					if let Some(jit_channel) = peer_state.outbound_channels_by_scid.get_mut(&scid) {
-						// TODO: Need to support MPP payments. If payment_amount_msat is known, needs to queue intercepted HTLCs in a map by payment_hash
-						//       LiquidityManager will need to be regularly polled so it can continually check if the payment amount has been received
-						//       and can release the payment or if the channel valid_until has expired and should be failed.
```
I take it you don't intend to handle timeouts/`valid_until` in this PR?

johncantrell97 replied:
Yeah, but we have issues logged in the repo for the timeouts so I felt it was safe to remove in-code "todo comments".

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -625,25 +683,17 @@ where
 									user_channel_id: scid as u128,
 								}));
 							}
+							Ok(None) => {}
```
Should we at least log our progress here?

johncantrell97 replied:
yes probably. we currently don't have access to a `Logger` in channel state or even the JitChannelManager right now.  Should we pass one around everywhere here?

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -675,13 +725,39 @@ where
+
+								let total_msat = htlcs
+									.iter()
+									.fold(0, |acc, htlc| acc + htlc.expected_outbound_amount_msat);
```
Same nit as above: lets use `sum` for this.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -675,13 +725,39 @@ where
+										htlc.expected_outbound_amount_msat - min_htlc_msat;
+									let amount_of_fee_to_take =
+										std::cmp::min(fee_msat, max_we_can_take);
+									let amount_to_forward_msat =
```
Let's not shadow `amt_to_forward_msat`, this is confusing.

johncantrell97 replied:
hm it's not _really_ shadowing it. I think I'll rename `amt_to_forward_msat` to `total_amt_to_forward_msat` and keep this one.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -675,13 +725,39 @@ where
+
+								for htlc in htlcs {
+									let max_we_can_take =
+										htlc.expected_outbound_amount_msat - min_htlc_msat;
```
I believe `next_outbound_htlc_minimum_msat` might change over time and would be influenced by `forward_intercepted_htlc`, so taking *exactly* this amount without updating would likely fail.

johncantrell97 replied:
yeah makes sense.  i could move the fetching of `next_outbound_htlc_minimum_msat` from the channel into the loop I guess so it happens before each htlc.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -675,13 +725,39 @@ where
+									let max_we_can_take =
+										htlc.expected_outbound_amount_msat - min_htlc_msat;
+									let amount_of_fee_to_take =
+										std::cmp::min(fee_msat, max_we_can_take);
```
I'm not quite getting why we'd take all fees from the first HTLCs but leave all the others untouched? Shouldn't we rather split it equally over all?

johncantrell97 replied:
well it's not necessarily just from the first:

in the PR I wrote:


> The spec does not specify how the LSP should extract the fee from the collection of HTLCs received as part of the MPP. In this pass I iterate over all parts and greedily take as much of the fee I can from each HTLC until the total fee amount has been accounted for. I could imagine another strategy that tries to minimize the number of HTLCs that are used to pay the fee (sorting from largest to smallest and performing same strategy but I'm not sure it matters.


We can't split the fee msat amount equally across all htlcs in case some htlcs are too small.  I guess we could try to do it proportionally from each HTLC? So calculate `fee_percentage = fee_msat / total_msat` and then take `htlc_amount_msat*fee_percentage` for each one? Feels like that could run into rounding issues and be slightly off.

I'm open to whatever strategy, the spec doesn't mention how to do it.

johncantrell97 commented:
I just pushed a new commit that handles _most_ of your review comments.  I simplified and renamed the OutboundJITChannelState to have the initial state immediately be `AwaitingPayment`.  

I have a couple of questions still:

1) Should I have JITChannelManager and the channel state objects take/hold a `Logger`? Right now they have no way to log outside of returning error messages that will get logged upstream.

2) The strategy for taking the fee from htlcs.  Currently I greedily take as much as I can from each htlc until I've collected enough for the total fee.  You suggested splitting it evenly, which would require doing so proportionally as a percentage of each htlc amount.  I think that would work but might run into rounding issues.

johncantrell97 commented:
1) We decided to punt on logging for now until the restructuring is settled.
2) I pushed a commit that takes the fee proportionally from each HTLC.  It rounds up to nearest msat for each HTLC and then takes the min(fee_remaining, proportional_fee) so that the last HTLC doesn't take more than whatever is left.  If we didn't take the min then the user would slightly overpay on fees (~1msat per htlc, roughly).

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -25,6 +24,7 @@ use lightning::routing::router::Router;
 use lightning::sign::{EntropySource, NodeSigner, SignerProvider};
 use lightning::util::errors::APIError;
 use lightning::util::logger::{Level, Logger};
+use lightning::{chain, log_debug};
```
nit: As we'll punt on logging, we shouldn't import the logging macro here.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -186,50 +192,86 @@ impl InboundJITChannel {
+					)
+				))?;
+
+				let amt_to_forward_msat = expected_payment_size_msat - opening_fee_msat;
```
Are we sure this can't underflow? Can we use `saturating_sub` here?

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -186,50 +192,86 @@ impl InboundJITChannel {
+						amt_to_forward_msat,
+					})
+				} else {
+					if supports_mpp {
```
Can we just make this an `if let Some(_)` and include a comment? Or at least move the `supports_mpp` down, as its defined way before we need it.

johncantrell97 replied:
Yeah either way I guess. I was trying to make it clear that Some(payment_size_msat) is how we determine if we support_mpp or not.

I think originally I was shadowing payment_size_msat so had moved the supports_mpp calculation up before unwrapping + shadowing it.  Since that's not the case I can move it down.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -258,34 +298,39 @@ impl OutboundJITChannelState {
 
 		match &self.state {
+			OutboundJITChannelState::AwaitingPayment { htlcs, payment_size_msat, .. } => {
+				// log that we received an htlc but are still awaiting payment
```
Let's include a `TODO` here so that we may find it again:
```suggestion
				// TODO: Should we log that we received an HTLC but are still awaiting payment?
```

johncantrell97 replied:
Sounds good, should probably create an issue too?

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -675,13 +712,40 @@ where
+								let mut fee_remaining_msat =
+									total_received_msat - total_amt_to_forward_msat;
+								let fee_percentage =
+									fee_remaining_msat as f64 / total_received_msat as f64;
```
Let's not introduce `f64`s here, they just make it hard to deal with fractional `msat`s and rounding errors.

johncantrell97 replied:
hm, how do you avoid it when working with percentages? i tried to avoid rounding errors below by always rounding up to nearest msat and then taking min of remaining amount

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -675,13 +712,40 @@ where
+										amount_to_forward_msat,
+									)?;
+
+									fee_remaining_msat -= actual_fee_amt_msat;
```
Are we sure we're not introducing some error here? How much would the LSP get underpaid here? Can we pull this logic out into a dedicated method and run some unit tests on it?

johncantrell97 replied:
hm? they can't get underpaid? I can still extract it and add tests to convince you :)

johncantrell97 replied:
ok pulled it out and added a test. not sure what other scenario to test exactly.

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -1237,3 +1281,70 @@ where
 	}
 }
+
+fn calculate_amount_to_forward(
```
nit: Can we include 'htlc' or 'per-htlc' in the method name to make it a bit clearer what it does?

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -1237,3 +1281,70 @@ where
+
+	#[test]
+	fn test_calculate_amount_to_forward() {
+		let htlcs = vec![
```
Could you add a TODO here saying `// TODO: Use proptest to generate random allocations` so we would be reminded to come back here?

Would be a good first test case to start implementing property-based testing. Happy to do that as a follow-up, tracking here: https://github.com/lightningdevkit/ldk-lsp-client/issues/45

tnull commented on src/jit_channel/channel_manager.rs:
```diff
@@ -1237,3 +1281,70 @@ where
+
+	let mut fee_remaining_msat = total_received_msat - total_amt_to_forward_msat;
+
+	let fee_percentage = fee_remaining_msat as f64 / total_received_msat as f64;
```
Still not convinced we really want/need these casting to `f64`.

johncantrell97 replied:
ok, I pushed a version using integer division but I'm way less confident in it.
